torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+      switch (j.type()) {
+        case json::value_t::number_integer:
+        case json::value_t::number_unsigned:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getI64IntegerAttr(j.template get<int64_t>())});
+          break;
+        case json::value_t::number_float:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getF32FloatAttr(j.template get<float>())});
+          break;
+        case json::value_t::boolean:
+          named_attrs.push_back({builder.getStringAttr(key),
+                                 builder.getBoolAttr(j.template get<bool>())});
+          break;
+        case json::value_t::string:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getStringAttr(j.template get<std::string>())});
+          break;
+        default:
+          // Ignored unrecognizable attr json
+          break;
+      }
+    }
+    return builder.getDictionaryAttr(named_attrs);
+  }
+
+  void BuildStableHLOCompositeOp(
+      mlir::Operation* op,
+      const llvm::DenseMap<const mlir::Operation*, size_t>& op_line_num) {
+    mlir::ModuleOp module_op = getOperation();
+    mlir::MLIRContext* context = &getContext();
+    mlir::OpBuilder builder(context);
+
+    std::unique_ptr<BoundaryMetadata> metadata = GetBoundaryMetadata(op);
+    if (metadata == nullptr || metadata->is_input) {
+      return;
+    }
+    const auto& output_metadata = *metadata;
+
+    llvm::SetVector<mlir::Operation*> scope_ops_setvec;
+    llvm::SetVector<std::pair<mlir::Value, int64_t>> arg_pos_setvec;
+    llvm::SmallVector<mlir::Operation*> processing({op});
+
+    // Reverse graph traversal: from boundary output op to boundary input op,
+    // global function arg, or stablehlo constant.
+    while (!processing.empty()) {
+      mlir::Operation* curr_op = processing.back();
+      processing.pop_back();
+      if (scope_ops_setvec.contains(curr_op)) {
+        continue;
+      }
+
+      if (auto curr_metadata_ptr = GetBoundaryMetadata(curr_op);
+          curr_metadata_ptr != nullptr) {
+        const auto& curr_metadata = *curr_metadata_ptr;
+        if (curr_metadata.is_input &&
+            curr_metadata.boundary_id() == output_metadata.boundary_id()) {
+          // Terminal condition: boundary input op.
+          arg_pos_setvec.insert({curr_op->getResult(0).dyn_cast<mlir::Value>(),
+                                 curr_metadata.pos});
+          continue;
+        }
+      }
+
+      scope_ops_setvec.insert(curr_op);
+      for (mlir::Value value : curr_op->getOperands()) {
+        mlir::Operation* def_op = value.getDefiningOp();
+        if (def_op == nullptr) {
+          // Terminal condition: Global function arg
+          arg_pos_setvec.insert({value, std::numeric_limits<int64_t>::max()});
+        } else if (llvm::isa<mlir::stablehlo::ConstantOp>(op)) {
+          // Terminal condition: constant
```

Comment:
Same question, should constants go through a `mark_begin` call?

Reply:
I think that not having _everything_ go through `mark` calls puts us at risk of corrupt abstractions. I.e. what does the following test case do?

```
x = mark_begin(x, "a")
x = x + 1
y = mark_begin(y, "b")
z = x + y
z = mark_end(z, "a")
z = z*2
z = mark_end(z, "b")
```

This has overlapping regions between separate abstractions.. I _think_ that one of the calls to begin would be consumed when creating an abstraction for `"a"`, and the `"b"` call may still be successful since it will trace back to an input argument or constant?

Reply:
`mark_begin` only accepts tensors, while the lowering can turn Python scalars that into StableHLO constant tensors.

The example is invalid: we have no plans to accept overlapped patterns. The behavior would be undefined or even wrong.

Reply:
The frontend in ODML SDK will make sure no overlapped patterns are marked in the long term.

Reply:
> mark_begin only accepts pytorch tensors, while the torch_xla

Would be good to settle on one thing, but even so take that suggestion as pseudo code, a hypothetical `torch_xla.mark_begin`.

> The behavior would be undefined or even wrong.

I guess my proposal is that the behavior should be "error"